django/core/cache/backends/redis.py

```diff
@@ -0,0 +1,239 @@
+        index = self._get_connection_pool_index(write)
+
+        params = self._parse_url(self._servers[index])
+        params.update(self._client_kwargs)
+
+        if self._pools[index] is None:
+            if self._servers[index].startswith('unix'):
+                self._pools[index] = self._pool_class(
+                    connection_class=redis.UnixDomainSocketConnection,
+                    path=self._servers[index][5:],
+                    **self._client_kwargs
+                )
+            else:
+                self._pools[index] = self._pool_class(**params)
+
+        return self._pools[index]
+
+    def get_client(self, key='', write=False):
+        pool = self._get_connection_pool(write)
+        return self._client(connection_pool=pool)
+
+    def get(self, key, default):
+        client = self.get_client(key)
+        value = client.get(key)
+        # If key does not exists, return None
+        if value is None:
+            return default
+        else:
+            return self._serializer.loads(value)
+
+    def add(self, key, value, timeout):
+        client = self.get_client(key)
+        return bool(client.set(key, self._serializer.dumps(value), ex=timeout, nx=True))
+
+    def set(self, key, value, timeout):
+        client = self.get_client(key, write=True)
+        client.set(key, self._serializer.dumps(value), ex=timeout)
+
+    def touch(self, key, timeout):
+        client = self.get_client(key, write=True)
+        if timeout is None:
+            return client.persist(key)
+        else:
+            return client.expire(key, timeout)
+
+    def delete(self, key):
+        client = self.get_client(key, write=True)
+        return bool(client.delete(key))
+
+    def clear(self):
+        client = self.get_client(None, write=True)
+        return client.flushall()
+
+    def get_many(self, keys):
+        client = self.get_client(None)
+        ret = client.mget(keys)
+        return {
+            k: self._serializer.loads(v) for k, v in zip(keys, ret) if v is not None
+        }
+
+    def set_many(self, data, timeout):
+        client = self.get_client(None)
+        serialized_data = {k: self._serializer.dumps(v) for k, v in data.items()}
+
+        client.mset(serialized_data)
+
+        for k, _ in data.items():
+            # Setting timeout for each as redis-py does not support timeout with mset
+            if timeout is None:
+                client.persist(k)
+            else:
+                client.expire(k, timeout)
+
+
+class RedisCache(BaseCache):
+    def __init__(self, server, params):
+        super().__init__(params)
+        if isinstance(server, str):
+            self._servers = re.split('[;,]', server)
+        else:
+            self._servers = server
+
+        self._options = params.get('OPTIONS') or {}
+        self._class = self._options.get('CLIENT', RedisCacheClient)
+
+        if isinstance(self._class, str):
+            self._class = import_string(self._class)
```

Comment:
As discussed earlier, I think that we should strip out the `CLIENT` bit for now to keep things simple. It is currently untested anyway and should be added if we decide to take this approach over providing a subclass of `RedisCache` with `self._class` overridden to a different client class. I suspect, however, that we should be able to do most things by passing options instead.
```suggestion
        self._class = RedisCacheClient
        self._options = params.get('OPTIONS') or {}
```

Reply:
So basically instead of having separate client class, we make the clients configurable using the `OPTIONS`, right?

Reply:
Yes. ISTM that it is worth trying to do it via options first and then we can always look to use a different client class if it becomes absolutely necessary. Let's start out with something simple.

The other benefit is that we (well, you!) are writing the client class and not linking to a third-party package. As such we can adjust it to fit our needs exactly. (When I wrote the `PyMemcacheCache` backend I even contributed changes to `pymemcache` to address some issues with compatibility as that was preferable to creating another custom client.)